src/Host/Client/Impl/Host/RemoteCredentialsDecorator.cs

```diff
@@ -41,24 +46,25 @@ internal class RemoteCredentialsDecorator : ICredentialsDecorator {
 
             await _mainThread.SwitchToAsync(cancellationToken);
 
-            try {
-                _credentials = await _securityService.GetUserCredentialsAsync(_authority, _workspaceName, cancellationToken);
-            } catch (Exception) {
-                token.Dispose();
-                throw;
+            lock (_credentialsLock) {
+                try {
+                    _credentials = _credentials ?? _securityService.GetUserCredentials(_authority, _workspaceName, cancellationToken);
+                    SecurityUtilities.WriteCredentials(_authority, _credentials);
```

Comment:
If we set _credentials to null upon invalidation then yes this has to be written under the lock.